pkg/controller/oam/applicationconfiguration/applicationconfiguration.go

```diff
@@ -0,0 +1,315 @@
+func WithApplicator(a WorkloadApplicator) ReconcilerOption {
+	return func(rc *Reconciler) {
+		rc.workloads = a
+	}
+}
+
+// WithGarbageCollector specifies how the Reconciler should garbage collect
+// workloads and traits when an ApplicationConfiguration is edited to remove
+// them.
+func WithGarbageCollector(gc GarbageCollector) ReconcilerOption {
+	return func(rc *Reconciler) {
+		rc.gc = gc
+	}
+}
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(l logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = l
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// NewReconciler returns a Reconciler that reconciles ApplicationConfigurations
+// by rendering and instantiating their Components and Traits.
+func NewReconciler(m ctrl.Manager, o ...ReconcilerOption) *Reconciler {
+	// NOTE(negz): We take a ctrl.Manager here despite only using its client for
+	// consistency with other NewReconciler functions, and to avoid needing to
+	// change the signature if we eventually need more from the manager (e.g its
+	// scheme).
+
+	r := &Reconciler{
+		client: m.GetClient(),
+		components: &components{
+			client:   m.GetClient(),
+			params:   ParameterResolveFn(resolve),
+			workload: ResourceRenderFn(renderWorkload),
+			trait:    ResourceRenderFn(renderTrait),
+		},
+		workloads: &workloads{
+			client: m.GetClient(),
+		},
+		gc:     GarbageCollectorFn(eligible),
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// NOTE(negz): We don't validate anything against their definitions at the
+// controller level. We assume this will be done by validating admission
+// webhooks.
+
+// Reconcile an OAM ApplicationConfigurations by rendering and instantiating its
+// Components and Traits.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	ac := &v1alpha2.ApplicationConfiguration{}
+	if err := r.client.Get(ctx, req.NamespacedName, ac); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetAppConfig)
+	}
+
+	log = log.WithValues("uid", ac.GetUID(), "version", ac.GetResourceVersion())
+
+	workloads, err := r.components.Render(ctx, ac)
+	if err != nil {
+		log.Debug("Cannot render components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotRenderComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errRenderComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
```

Comment:
I wonder why is the error "errUpdateAppConfigStatus" and is it really updated to the applicationConfig's status?